File: packages/core/src/utils/run-log.ts
/**
 * R26 调用级运行遥测（v5 五轮规划 P0，403 号）。
 *
 * 每次 LLM 调用（链内一轮 = 一条）追加元数据记录到进程级环形缓冲：
 * agent / model / 耗时 ms / 成败 / 链内序号 / 轮次 / 是否接管 / 错误类。
 * **只记元数据，绝不记 prompt/正文/错误原文**（v5 隐私红线；错误原文
 * 走 inkos.log 的 R25 既有通道）。
 *
 * 双端：`engine-rs/src/utils/run_log.rs` 1:1；共享 golden
 * `golden/run-log-vectors.json`（TS `golden-run-log.test.ts` /
 * RS `tests/golden_run_log_diff.rs`）。
 */

/** 默认环形缓冲容量（v5 规格：200 条）。 */
export const DEFAULT_RUN_LOG_CAPACITY = 200;

export type RunLogErrorKind = "transient" | "fatal";

export interface RunLogEntry {
  /** ISO 8601 UTC 时间戳（记录时刻）。 */
  readonly ts: string;
  /** agent 名 / 调用标签。 */
  readonly agent: string;
  readonly model: string;
  /** 调用耗时（毫秒，整数）。 */
  readonly durationMs: number;
  readonly ok: boolean;
  /** 链内序号（0 = primary）。 */
  readonly attemptIndex: number;
  /** 轮次（1 起）。 */
  readonly round: number;
  /** R25 联动：非 primary 首轮命中即接管切换。 */
  readonly tookOver: boolean;
  readonly errorKind: RunLogErrorKind | null;
}

export interface RunLogSnapshot {
  /** 缓冲内现存记录（旧 → 新）。 */
  readonly entries: ReadonlyArray<RunLogEntry>;
  /** 累计追加数（含被逐出的旧记录）。 */
  readonly totalAppended: number;
}

export interface RunLogProjection {
  /** 累计调用数（含被逐出）。 */
  readonly total: number;
  /** 缓冲内现存条数。 */
  readonly kept: number;
  /** 现存记录中接管成功的次数（R25 联动展示）。 */
  readonly tookOverCount: number;
  /** 现存记录中失败次数。 */
  readonly failureCount: number;
  /** 最新在后（时间升序，环形缓冲自然序）。 */
  readonly entries: ReadonlyArray<RunLogEntry>;
}

export class RunLogBuffer {
  private items: RunLogEntry[] = [];
  private totalAppended = 0;

  constructor(readonly capacity: number = DEFAULT_RUN_LOG_CAPACITY) {}

  append(entry: RunLogEntry): void {
    this.totalAppended += 1;
    this.items.push(entry);
    if (this.items.length > this.capacity) {
      this.items.splice(0, this.items.length - this.capacity);
    }
  }

  snapshot(): RunLogSnapshot {
    return { entries: [...this.items], totalAppended: this.totalAppended };
  }
}

/**
 * 只读投影：聚合计数 + 可选截尾（limit = 返回最近 limit 条，仍按时间升序）。
 * limit 非正或缺省 → 全量。
 */
export function projectRunLog(
  snapshot: RunLogSnapshot,
  limit?: number,
): RunLogProjection {
  const entries = limit !== undefined && limit > 0
    ? snapshot.entries.slice(Math.max(0, snapshot.entries.length - limit))
    : snapshot.entries;
  return {
    total: snapshot.totalAppended,
    kept: snapshot.entries.length,
    tookOverCount: snapshot.entries.filter((entry) => entry.tookOver).length,
    failureCount: snapshot.entries.filter((entry) => !entry.ok).length,
    entries,
  };
}

/** 进程级单例（studio server / engine 各自进程内）。 */
export const globalRunLog = new RunLogBuffer();

File: engine-rs/src/utils/run_log.rs
//! R26 调用级运行遥测（v5 五轮规划 P0，403 号）。
//!
//! TS 真源：`packages/core/src/utils/run-log.ts`。每次 LLM 调用（链内一轮）
//! 追加元数据到进程级环形缓冲——**只记元数据，绝不记 prompt/正文/错误原文**
//! （v5 隐私红线；错误原文走 inkos.log 的 R25 既有通道）。
//! 共享向量 `packages/core/src/__tests__/golden/run-log-vectors.json`
//! 差分：`tests/golden_run_log_diff.rs`。

use std::sync::Mutex;
use std::sync::OnceLock;

pub const DEFAULT_RUN_LOG_CAPACITY: usize = 200;

#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RunLogEntry {
    /// ISO 8601 UTC 时间戳（记录时刻）。
    pub ts: String,
    /// agent 名 / 调用标签。
    pub agent: String,
    pub model: String,
    /// 调用耗时（毫秒，整数）。
    pub duration_ms: i64,
    pub ok: bool,
    /// 链内序号（0 = primary）。
    pub attempt_index: i64,
    /// 轮次（1 起）。
    pub round: i64,
    /// R25 联动：非 primary 首轮命中即接管切换。
    pub took_over: bool,
    pub error_kind: Option<String>,
}

#[derive(Debug, Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RunLogSnapshot {
    /// 缓冲内现存记录（旧 → 新）。
    pub entries: Vec<RunLogEntry>,
    /// 累计追加数（含被逐出的旧记录）。
    pub total_appended: i64,
}

pub struct RunLogBuffer {
    items: Vec<RunLogEntry>,
    total_appended: i64,
    capacity: usize,
}

impl RunLogBuffer {
    pub fn new(capacity: usize) -> Self {
        Self { items: Vec::new(), total_appended: 0, capacity: capacity.max(1) }
    }

    pub fn append(&mut self, entry: RunLogEntry) {
        self.total_appended += 1;
        self.items.push(entry);
        if self.items.len() > self.capacity {
            let excess = self.items.len() - self.capacity;
            self.items.drain(..excess);
        }
    }

    pub fn snapshot(&self) -> RunLogSnapshot {
        RunLogSnapshot { entries: self.items.clone(), total_appended: self.total_appended }
    }
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct RunLogProjection {
    /// 累计调用数（含被逐出）。
    pub total: i64,
    /// 缓冲内现存条数。
    pub kept: i64,
    /// 现存记录中接管成功的次数（R25 联动展示）。
    pub took_over_count: i64,
    /// 现存记录中失败次数。
    pub failure_count: i64,
    /// 最新在后（时间升序，环形缓冲自然序）。
    pub entries: Vec<RunLogEntry>,
}

/// 只读投影：聚合计数 + 可选截尾（limit = 返回最近 limit 条，仍按时间升序）。
/// limit 非正或缺省 → 全量。
pub fn project_run_log(snapshot: &RunLogSnapshot, limit: Option<i64>) -> RunLogProjection {
    let entries = match limit {
        Some(limit) if limit > 0 => {
            let skip = (snapshot.entries.len() as i64 - limit).max(0) as usize;
            snapshot.entries[skip.min(snapshot.entries.len())..].to_vec()
        }
        _ => snapshot.entries.clone(),
    };
    RunLogProjection {
        total: snapshot.total_appended,
        kept: snapshot.entries.len() as i64,
        took_over_count: snapshot.entries.iter().filter(|entry| entry.took_over).count() as i64,
        failure_count: snapshot.entries.iter().filter(|entry| !entry.ok).count() as i64,
        entries,
    }
}

/// 进程级单例（engine 进程内）。
pub fn global_run_log() -> &'static Mutex<RunLogBuffer> {
    static GLOBAL: OnceLock<Mutex<RunLogBuffer>> = OnceLock::new();
    GLOBAL.get_or_init(|| Mutex::new(RunLogBuffer::new(DEFAULT_RUN_LOG_CAPACITY)))
}

/// 便捷记录：构造条目（时间戳内部生成）并追加入全局缓冲。
pub fn record_run_log(entry: RunLogEntry) {
    if let Ok(mut buffer) = global_run_log().lock() {
        buffer.append(entry);
    }
}
